File: c++/bubble_sort/bubble_sort.cpp
#include <iostream>
#include <stdlib.h>

using namespace std;

void fillArray(int* ary, int max, int val_min_inclusive, int val_max_exclusive)
{
  for (int i = 0; i < max; ++i) {
    ary[i] = val_min_inclusive + (rand() % val_max_exclusive);
  }
}

void printArray(int* ary, int max)
{
  int i = 0;
  for (; i < max - 1; ++i) {
    cout << ary[i] << " ";
  }
  cout << ary[i] << endl;
}

void bubbleSort(int* ary, int max)
{
  for (int outer = max - 1; outer > 0; outer--) {
    for(int inner = 0; inner < outer; inner++) {
      if (ary[inner] > ary[inner + 1]) {
        swap(ary[inner], ary[inner + 1]);
        // int temp = ary[inner + 1];
        // ary[inner + 1] = ary[inner];
        // ary[inner] = temp;
      }
    }
  }
}

int main()
{
  srand(time(NULL));

  int ary_min_count = 2;
  int ary_max_count = 10;
  for (int i = ary_min_count; i <= ary_max_count; i++) {
    int ary[i] = {};
    int* ary_ptr = &ary[0];

    fillArray(ary_ptr, i, -1000, 2000);
    printArray(ary_ptr, i);
    bubbleSort(ary_ptr, i);
    printArray(ary_ptr, i);
    cout << endl;
  }
  return 0;
}

File: python/bubble_sort/bubble_sort.py
import random

aryLen = 20
ary = range(1,aryLen+1)

def knuthShuffle(ary):
  aryLen = len(ary)
  for i in range(1, aryLen+1):
    j = aryLen - i
    if j > 0:
      swapI = random.randint(0, j-1)
    else:
      swapI = 0
    temp = ary[swapI]
    ary[swapI] = ary[j]
    ary[j] = temp
  pass

def bubbleSort(ary):
  aryLen = len(ary)
  for i in range(1, aryLen + 1):
    outer = aryLen - i
    for inner in range(0, outer):
      if (ary[inner] > ary[inner + 1]):
        temp = ary[inner]
        ary[inner] = ary[inner + 1]
        ary[inner + 1] = temp
  pass

print ary
knuthShuffle(ary)
print ary
bubbleSort(ary)
print ary